File: tests/authoring/test_publisher.py
from dataclasses import replace
from datetime import date

import pytest

from log_migration.authoring.models import PageProblem, PublishError, Redirect, SaveRequest
from log_migration.authoring.repository import RepositorySnapshot

from .support import new_date_request, repository_for


def test_build_excludes_draft_body_and_includes_referenced_empty_page(tmp_path):
    repository = repository_for(tmp_path)
    published = repository.save_draft(new_date_request(date(2026, 1, 1), "公開本文\n\n[[page-a]]"))
    target = repository.find_route("page-a")
    assert target is not None
    repository.save_draft(SaveRequest(page_type="named", page_id=target.id, body="下書き本文"))
    destination = tmp_path / "next-site"

    publisher(tmp_path).build(published_snapshot(repository.refresh(), published.id), destination)

    assert "公開本文" in (destination / "2026-01-01" / "index.html").read_text()
    placeholder = (destination / "page-a" / "index.html").read_text()
    assert "page-a" in placeholder
    assert "まだ内容がありません" in placeholder
    assert "下書き本文" not in "".join(path.read_text() for path in destination.rglob("*.html"))


def test_build_includes_only_published_backlinks(tmp_path):
    repository = repository_for(tmp_path)
    published = repository.save_draft(new_date_request(date(2026, 1, 1), "[[page-a]]"))
    draft = repository.save_draft(SaveRequest(page_type="named", name="draft-source", body="[[page-a]]"))
    destination = tmp_path / "next-site"

    publisher(tmp_path).build(published_snapshot(repository.refresh(), published.id), destination)

    page = (destination / "page-a" / "index.html").read_text()
    assert "2026-01-01" in page
    assert draft.display_title not in page
    assert not (destination / "draft-source").exists()


def test_build_writes_redirect_only_for_published_named_page(tmp_path):
    repository = repository_for(tmp_path)
    published = repository.save_draft(SaveRequest(page_type="named", name="new-name", body="公開本文"))
    repository.save_draft(SaveRequest(page_type="named", name="draft-name", body="下書き本文"))
    snapshot = replace(
        published_snapshot(repository.refresh(), published.id),
        redirects=(
            Redirect(old_route="old-name", new_route="new-name"),
            Redirect(old_route="old-draft", new_route="draft-name"),
        ),
    )
    destination = tmp_path / "next-site"

    publisher(tmp_path).build(snapshot, destination)

    redirect = (destination / "old-name" / "index.html").read_text()
    assert 'url=/new-name' in redirect
    assert not (destination / "old-draft").exists()


def test_build_flattens_redirect_chain_to_final_published_route(tmp_path):
    repository = repository_for(tmp_path)
    page = repository.save_draft(SaveRequest(page_type="named", name="page-c", body="本文"))
    snapshot = replace(
        published_snapshot(repository.refresh(), page.id),
        redirects=(
            Redirect(old_route="page-a", new_route="page-b"),
            Redirect(old_route="page-b", new_route="page-c"),
        ),
    )
    destination = tmp_path / "next-site"

    publisher(tmp_path).build(snapshot, destination)

    page_a = (destination / "page-a" / "index.html").read_text(encoding="utf-8")
    page_b = (destination / "page-b" / "index.html").read_text(encoding="utf-8")
    assert 'url=/page-c' in page_a
    assert 'url=/page-c' in page_b
    assert 'url=/page-b' not in page_a


def test_build_rejects_problem_required_by_public_candidate(tmp_path):
    repository = repository_for(tmp_path)
    published = repository.save_draft(new_date_request(date(2026, 1, 1), "[[page-a]]"))
    snapshot = replace(
        published_snapshot(repository.refresh(), published.id),
        problems=(PageProblem(tmp_path / "content" / "page-a.md", "broken frontmatter"),),
    )

    with pytest.raises(PublishError, match="page-a"):
        publisher(tmp_path).build(snapshot, tmp_path / "next-site")


def test_build_ignores_problem_unrelated_to_public_candidate(tmp_path):
    repository = repository_for(tmp_path)
    published = repository.save_draft(new_date_request(date(2026, 1, 1), "公開本文"))
    snapshot = replace(
        published_snapshot(repository.refresh(), published.id),
        problems=(PageProblem(tmp_path / "content" / "draft-only.md", "broken frontmatter"),),
    )
    destination = tmp_path / "next-site"

    publisher(tmp_path).build(snapshot, destination)

    assert "公開本文" in (destination / "2026-01-01" / "index.html").read_text()


def test_build_url_encodes_named_page_route(tmp_path):
    repository = repository_for(tmp_path)
    page = repository.save_draft(SaveRequest(page_type="named", name="page a", body="公開本文"))
    destination = tmp_path / "next-site"

    publisher(tmp_path).build(published_snapshot(repository.refresh(), page.id), destination)

    assert (destination / "page%20a" / "index.html").exists()
    assert not (destination / "page a").exists()


def test_publish_rejects_route_that_escapes_staging_destination(tmp_path):
    repository = repository_for(tmp_path)
    page = repository.save_draft(SaveRequest(page_type="named", name="..", body="公開本文"))
    current_site = tmp_path / "site"
    current_site.mkdir()
    (current_site / "marker").write_text("before", encoding="utf-8")

    with pytest.raises(PublishError, match="escapes destination"):
        publisher(tmp_path).publish(published_snapshot(repository.refresh(), page.id))

    assert (current_site / "marker").read_text(encoding="utf-8") == "before"
    assert not (tmp_path / "index.html").exists()
    assert not tuple(tmp_path.glob("site.staging-*"))


def test_build_rejects_redirect_that_overwrites_current_public_route(tmp_path):
    repository = repository_for(tmp_path)
    page = repository.save_draft(SaveRequest(page_type="named", name="current", body="公開本文"))
    snapshot = replace(
        published_snapshot(repository.refresh(), page.id),
        redirects=(Redirect(old_route="current", new_route="current"),),
    )

    with pytest.raises(PublishError, match="redirect.*current"):
        publisher(tmp_path).build(snapshot, tmp_path / "next-site")


def test_publish_keeps_existing_site_when_swap_fails(tmp_path, monkeypatch):
    repository = repository_for(tmp_path)
    published = repository.save_draft(new_date_request(date(2026, 1, 1), "公開本文"))
    current_site = tmp_path / "site"
    current_site.mkdir()
    (current_site / "marker").write_text("before", encoding="utf-8")
    static_publisher = publisher(tmp_path)

    import log_migration.authoring.publisher as publisher_module

    original_replace = publisher_module.os.replace

    def fail_staging_swap(source, destination):
        if source.name.startswith("site.staging-") and destination == current_site:
            raise OSError("simulated swap failure")
        original_replace(source, destination)

    monkeypatch.setattr(publisher_module.os, "replace", fail_staging_swap)

    with pytest.raises(PublishError, match="simulated swap failure"):
        static_publisher.publish(published_snapshot(repository.refresh(), published.id))

    assert (current_site / "marker").read_text(encoding="utf-8") == "before"
    assert not tuple(tmp_path.glob("site.staging-*"))
    assert not tuple(tmp_path.glob("site.previous-*"))


def test_publish_does_not_create_site_when_build_fails(tmp_path):
    repository = repository_for(tmp_path)
    published = repository.save_draft(new_date_request(date(2026, 1, 1), "valid"))
    snapshot = published_snapshot(repository.refresh(), published.id)
    broken = replace(snapshot.pages[0], body="[[bad/name]]")
    snapshot = replace(snapshot, pages=(broken,))

    with pytest.raises(PublishError, match="forbidden"):
        publisher(tmp_path).publish(snapshot)

    assert not (tmp_path / "site").exists()
    assert not tuple(tmp_path.glob("site.staging-*"))


def publisher(tmp_path):
    from log_migration.authoring.publisher import StaticPublisher

    return StaticPublisher(tmp_path / "site")


def published_snapshot(snapshot: RepositorySnapshot, page_id: str) -> RepositorySnapshot:
    return replace(
        snapshot,
        pages=tuple(
            replace(page, status="published", published_at=page.updated_at)
            if page.id == page_id
            else page
            for page in snapshot.pages
        ),
    )

File: test/authoring/test_publisher.rb
# frozen_string_literal: true

require_relative "../test_helper"
require_relative "../../lib/weblog_authoring/publisher"

require "fileutils"
require "tmpdir"

class TestPublisher < Minitest::Test
  FIXED_TIME = Time.iso8601("2026-01-01T00:00:00+09:00")

  def test_release_manifest_round_trips_and_preserves_redirects
    page = published_named_page("page-a", body: "公開本文")
    snapshot = WeblogAuthoring::ReleaseSnapshot.new(
      pages: [page],
      redirects: [
        WeblogAuthoring::Redirect.new(old_route: "old-a", new_route: "old-b"),
        WeblogAuthoring::Redirect.new(old_route: "old-b", new_route: "page-a")
      ],
      published_at: FIXED_TIME
    )
    manifest = WeblogAuthoring::ReleaseManifest.new(tmpdir.join("content/.authoring-release.json"))

    serialized = manifest.serialize(snapshot)
    manifest.path.dirname.mkpath
    manifest.path.write(serialized, encoding: "UTF-8")
    loaded = manifest.load

    assert_equal snapshot.pages, loaded.pages
    assert_equal(
      [
        WeblogAuthoring::Redirect.new(old_route: "old-a", new_route: "page-a"),
        WeblogAuthoring::Redirect.new(old_route: "old-b", new_route: "page-a")
      ],
      loaded.redirects
    )
    assert_equal snapshot.published_at, loaded.published_at
  end

  def test_release_manifest_rejects_malformed_metadata
    manifest = WeblogAuthoring::ReleaseManifest.new(tmpdir.join("content/.authoring-release.json"))
    manifest.path.dirname.mkpath
    manifest.path.write("{\"version\":1,\"pages\":[],\"redirects\":[]}\n", encoding: "UTF-8")

    error = assert_raises(WeblogAuthoring::PublishError) { manifest.load }

    assert_includes error.message, "missing root key"
  end

  def test_release_manifest_rejects_missing_file
    manifest = WeblogAuthoring::ReleaseManifest.new(tmpdir.join("content/.authoring-release.json"))

    error = assert_raises(WeblogAuthoring::PublishError) { manifest.load }

    assert_includes error.message, "missing"
  end

  def test_build_uses_release_snapshot_body_for_existing_published_page
    published = published_date_page("2026-01-01", body: "未公開編集\n\n[[page-a]]")
    placeholder_source = draft_named_page("page-a", body: "下書き本文")
    release_snapshot = WeblogAuthoring::ReleaseSnapshot.new(
      pages: [published_date_page("2026-01-01", body: "旧公開本文")],
      redirects: [],
      published_at: FIXED_TIME
    )

    result = publisher.build(
      snapshot_for(pages: [published, placeholder_source]),
      tmpdir.join("site.staging"),
      release_snapshot:
    )

    public_html = tmpdir.join("site.staging/2026-01-01/index.html").read(encoding: "UTF-8")
    assert_includes public_html, "旧公開本文"
    refute_includes public_html, "未公開編集"
    assert_equal ["2026-01-01"], result.public_routes.sort
    assert_equal "未公開編集\n\n[[page-a]]", result.release_candidate.pages.fetch(0).body
    assert_equal FIXED_TIME, result.release_candidate.pages.fetch(0).published_at
  end

  def test_build_uses_current_body_and_placeholder_for_first_publish
    published = published_date_page("2026-01-01", body: "公開本文\n\n[[page-a]]")
    placeholder_source = draft_named_page("page-a", body: "下書き本文")

    result = publisher.build(
      snapshot_for(pages: [published, placeholder_source]),
      tmpdir.join("site.staging"),
      release_snapshot: WeblogAuthoring::ReleaseSnapshot.new
    )

    public_html = tmpdir.join("site.staging/2026-01-01/index.html").read(encoding: "UTF-8")
    placeholder_html = tmpdir.join("site.staging/page-a/index.html").read(encoding: "UTF-8")

    assert_includes public_html, "公開本文"
    assert_includes placeholder_html, "まだ内容がありません"
    assert_includes placeholder_html, "2026-01-01"
    refute_includes placeholder_html, "下書き本文"
    assert_equal ["2026-01-01", "page-a"], result.public_routes.sort
  end

  def test_build_published_rename_uses_released_body_at_new_route_and_redirects_old_route
    current_page = published_named_page("new-name", id: "renamed-page", body: "未公開の新本文")
    release_snapshot = WeblogAuthoring::ReleaseSnapshot.new(
      pages: [published_named_page("old-name", id: "renamed-page", body: "旧公開本文")],
      redirects: [],
      published_at: FIXED_TIME
    )

    result = publisher.build(
      snapshot_for(
        pages: [current_page],
        redirects: [WeblogAuthoring::Redirect.new(old_route: "old-name", new_route: "new-name")]
      ),
      tmpdir.join("site.staging"),
      release_snapshot:
    )

    new_route_html = tmpdir.join("site.staging/new-name/index.html").read(encoding: "UTF-8")
    old_route_html = tmpdir.join("site.staging/old-name/index.html").read(encoding: "UTF-8")

    assert_includes new_route_html, "旧公開本文"
    refute_includes new_route_html, "未公開の新本文"
    assert_includes old_route_html, "url=/new-name"
    assert_equal ["new-name", "old-name"], result.public_routes.sort
    assert_equal "未公開の新本文", result.release_candidate.pages.fetch(0).body
    assert_equal "new-name", result.release_candidate.pages.fetch(0).route
  end

  def test_build_does_not_output_draft_only_reference
    draft_source = draft_named_page("draft-source", body: "[[page-a]]")

    result = publisher.build(
      snapshot_for(pages: [draft_source]),
      tmpdir.join("site.staging"),
      release_snapshot: WeblogAuthoring::ReleaseSnapshot.new
    )

    assert_equal [], result.public_routes
    refute tmpdir.join("site.staging/page-a/index.html").exist?
  end

  def test_build_flattens_redirect_chain_and_skips_draft_only_redirect
    published = published_named_page("page-c")
    current_redirects = [
      WeblogAuthoring::Redirect.new(old_route: "page-b", new_route: "page-c"),
      WeblogAuthoring::Redirect.new(old_route: "draft-old", new_route: "draft-page")
    ]
    release_snapshot = WeblogAuthoring::ReleaseSnapshot.new(
      pages: [],
      redirects: [WeblogAuthoring::Redirect.new(old_route: "page-a", new_route: "page-b")],
      published_at: FIXED_TIME
    )

    result = publisher.build(
      snapshot_for(pages: [published], redirects: current_redirects),
      tmpdir.join("site.staging"),
      release_snapshot:
    )

    assert_includes tmpdir.join("site.staging/page-a/index.html").read(encoding: "UTF-8"), "url=/page-c"
    assert_includes tmpdir.join("site.staging/page-b/index.html").read(encoding: "UTF-8"), "url=/page-c"
    refute tmpdir.join("site.staging/draft-old/index.html").exist?
    assert_equal ["page-a", "page-b"], result.release_candidate.redirects.map(&:old_route).sort
  end

  def test_build_rejects_redirect_cycles_and_collisions
    published = published_named_page("page-c")
    cycle_snapshot = snapshot_for(
      pages: [published],
      redirects: [
        WeblogAuthoring::Redirect.new(old_route: "page-a", new_route: "page-b"),
        WeblogAuthoring::Redirect.new(old_route: "page-b", new_route: "page-a")
      ]
    )

    cycle_error = assert_raises(WeblogAuthoring::PublishError) do
      publisher.build(cycle_snapshot, tmpdir.join("cycle"), release_snapshot: WeblogAuthoring::ReleaseSnapshot.new)
    end
    assert_includes cycle_error.message, "cycle"

    collision_snapshot = snapshot_for(
      pages: [published, published_named_page("page-d")],
      redirects: [WeblogAuthoring::Redirect.new(old_route: "page-c", new_route: "page-d")]
    )

    collision_error = assert_raises(WeblogAuthoring::PublishError) do
      publisher.build(collision_snapshot, tmpdir.join("collision"), release_snapshot: WeblogAuthoring::ReleaseSnapshot.new)
    end
    assert_includes collision_error.message, "collides"
  end

  def test_build_rejects_public_source_problem
    published = published_named_page("page-a")
    problem = WeblogAuthoring::PageProblem.new(path: tmpdir.join("content/page-a.md"), detail: "broken frontmatter")

    error = assert_raises(WeblogAuthoring::PublishError) do
      publisher.build(
        snapshot_for(pages: [published], problems: [problem]),
        tmpdir.join("site.staging"),
        release_snapshot: WeblogAuthoring::ReleaseSnapshot.new
      )
    end

    assert_includes error.message, "page-a"
  end

  def test_build_ignores_draft_only_problem
    published = published_named_page("page-a")
    problem = WeblogAuthoring::PageProblem.new(path: tmpdir.join("content/draft-page.md"), detail: "broken frontmatter")

    result = publisher.build(
      snapshot_for(pages: [published], problems: [problem]),
      tmpdir.join("site.staging"),
      release_snapshot: WeblogAuthoring::ReleaseSnapshot.new
    )

    assert_equal ["page-a"], result.public_routes
    assert tmpdir.join("site.staging/page-a/index.html").exist?
  end

  def test_build_uses_case_distinguishing_output_route_encoding
    published = published_named_page("Page-A")

    result = publisher.build(
      snapshot_for(pages: [published]),
      tmpdir.join("site.staging"),
      release_snapshot: WeblogAuthoring::ReleaseSnapshot.new
    )

    assert_equal ["Page-A"], result.public_routes
    assert tmpdir.join("site.staging/%50age-%41/index.html").exist?
    refute tmpdir.join("site.staging/Page-A/index.html").exist?
  end

  def test_publish_keeps_existing_site_and_manifest_when_swap_fails
    published = published_named_page("page-a", body: "公開本文")
    site_dir = tmpdir.join("site")
    site_dir.mkpath
    site_dir.join("marker").write("before", encoding: "UTF-8")
    manifest_path = tmpdir.join("content/.authoring-release.json")
    manifest_path.dirname.mkpath
    manifest_path.write("before", encoding: "UTF-8")
    original_rename = File.method(:rename)
    publisher = publisher_for(site_dir)

    File.singleton_class.send(:define_method, :rename) do |source, destination|
      if Pathname(destination) == site_dir && Pathname(source).basename.to_s.start_with?("site.staging-")
        raise IOError, "simulated swap failure"
      end

      original_rename.call(source, destination)
    end

    error = assert_raises(WeblogAuthoring::PublishError) do
      publisher.publish(snapshot_for(pages: [published]), release_snapshot: WeblogAuthoring::ReleaseSnapshot.new)
    end

    assert_includes error.message, "simulated swap failure"
    assert_equal "before", site_dir.join("marker").read(encoding: "UTF-8")
    assert_equal "before", manifest_path.read(encoding: "UTF-8")
    assert_empty tmpdir.glob("site.staging-*")
    assert_empty tmpdir.glob("site.previous-*")
  ensure
    File.singleton_class.send(:define_method, :rename, original_rename) if defined?(original_rename) && original_rename
  end

  private

  def publisher
    @publisher ||= publisher_for(tmpdir.join("site"))
  end

  def publisher_for(path)
    WeblogAuthoring::StaticPublisher.new(path)
  end

  def snapshot_for(pages:, problems: [], redirects: [])
    WeblogAuthoring::RepositorySnapshot.new(pages:, problems:, redirects:)
  end

  def published_date_page(date_string, body: "")
    page_date = Date.iso8601(date_string)
    WeblogAuthoring::PageDocument.new(
      id: "date-#{date_string}",
      page_type: "date",
      name: nil,
      page_date:,
      title: nil,
      status: "published",
      created_at: FIXED_TIME,
      updated_at: FIXED_TIME,
      published_at: FIXED_TIME,
      path: tmpdir.join("content/#{date_string}.md"),
      body:,
      links: WeblogAuthoring.extract_wiki_links(body)
    )
  end

  def published_named_page(name, id: nil, body: "公開本文")
    WeblogAuthoring::PageDocument.new(
      id: id || "page-#{name}",
      page_type: "named",
      name:,
      page_date: nil,
      title: nil,
      status: "published",
      created_at: FIXED_TIME,
      updated_at: FIXED_TIME,
      published_at: FIXED_TIME,
      path: tmpdir.join("content/#{WeblogAuthoring.encoded_page_name(name)}.md"),
      body:,
      links: WeblogAuthoring.extract_wiki_links(body)
    )
  end

  def draft_named_page(name, body: "")
    WeblogAuthoring::PageDocument.new(
      id: "draft-#{name}",
      page_type: "named",
      name:,
      page_date: nil,
      title: nil,
      status: "draft",
      created_at: FIXED_TIME,
      updated_at: FIXED_TIME,
      published_at: nil,
      path: tmpdir.join("content/#{WeblogAuthoring.encoded_page_name(name)}.md"),
      body:,
      links: WeblogAuthoring.extract_wiki_links(body)
    )
  end

  def tmpdir
    @tmpdir ||= Pathname(Dir.mktmpdir("weblog-authoring-publisher"))
  end
end
